### PR TITLE
Functional: add toArray

### PR DESCRIPTION
**Pull request by ikarienator**
We shall let the user of `MonoidalReducer` to decide if the state is linear typed. Mutability should not get in people's way if it is not helping.

### CODE REVIEW

**Comment by ikarienator** [comment]
In fact if the empty method is not called every time, the current codegen shall fail.

**Comment by michaelficarra** [comment]
I do not like this at all. The identity is an element, not a function. If you want to mutate it, clone it.

**Comment by ikarienator** [comment]
That's what `Monoid.identity` should control. In our library we cannot make false assumptions like that.

**Comment by ikarienator** [comment]
It is also an unnecessary restriction to disallow linearly typed objects. If the Monoid.identity returns constant, it will be inlined by JVM.

**Comment by ikarienator** [comment]
If you stick to your reference transparency argument, the change has no effect at all. But for people who uses mutations in their systems (like 99.9% Java programmers), it is unacceptable to arbitrarily cache things they may not want you to cache.

**Comment by ikarienator** [comment]
Let me take that out and leave alone the `ConcatList.toArray` part.

**Comment by ikarienator** [comment]
Continue discussion on #43

File: src/main/java/com/shapesecurity/functional/data/ConcatList.java [hunk 1]
```diff
@@ -83,6 +85,18 @@ protected ConcatList(int length) {
   @NotNull
   public abstract Maybe<ConcatList<T>> update(int index, @NotNull T element);
 
+  @NotNull
+  public final T[] toArray(T[] array) {
+    @SuppressWarnings("unchecked")
+    T[] target = (array.length != this.length) ?
```
Comment (michaelficarra) [inline src/main/java/com/shapesecurity/functional/data/ConcatList.java]: Passing in an array with the wrong length should throw an error, not be glossed over.

Comment (ikarienator) [inline src/main/java/com/shapesecurity/functional/data/ConcatList.java]: That's not what it is usually designed. The array used as a carrier of the runtime type information of `T`.

**Pull request merged**